Add new `FeedbackDataset` class using `/api/v1` endpoints

alvarobartt opened the pull request:
## Description

This PR addresses the creation of a new `FeedbackDataset` class to handle the new datasets provided in Argilla. These datasets define some fields and questions for every dataset. The fields are the inputs, while the questions are the "annotations" which in this case correspond to questions over the fields that can either be numerical or textual, so that the annotator has to answer those questions in order to label the data.

This PR closes #2758, and #2764

**Type of change**

- [X] New feature (non-breaking change which adds functionality)

**How Has This Been Tested**

- [X] Included unit tests for `/api/v1/datasets` endpoint wrappers
- [X] Included unit tests for `rg.FeedbackDataset` class and any other related method/function

**Checklist**

- [X] I have merged the original branch into my forked branch
- [X] I added relevant documentation
- [X] follows the style guidelines of this project
- [X] I did a self-review of my code
- [X] I made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added relevant notes to the CHANGELOG.md file (See https://keepachangelog.com/)

## Example

```python
import argilla as rg

rg.init(
    api_url="<ARGILLA_URL_GOES_HERE>",
    api_key="<ARGILLA_API_KEY_GOES_HERE>",
)

dataset = rg.FeedbackDataset(
  guidelines="...",
  fields=[
    rg.TextField(name="prompt", required=True),
    rg.TextField(name="context", required=False),
    rg.TextField(name="completion", required=True),
  ],
  questions=[
    rg.RatingQuestion(name="prompt-quality", title="Assess the quality of the prompt", required=True),
    rg.TextQuestion(name="prompt-edit", title="Edit the prompt if applicable", required=False),
  ]
)

dataset.add_records(records=[
  rg.FeedbackRecord(
    fields={"prompt": ..., "context": ..., "completion": ...}
  ),
])
print(dataset.records[0])
dataset.push_to_argilla(name="my-dataset", workspace="my-workspace")
```

alvarobartt committed:
Add `FeedbackDataset` schema

alvarobartt committed:
Add `list_datasets` function

alvarobartt committed:
Merge branch 'feat/2615-instructions-task' of https://github.com/argilla-io/argilla into feat/list-datasets-client

alvarobartt committed:
Fix `FeedbackDataset` schema

alvarobartt committed:
Add docstrings and type-hints to `list_datasets`

alvarobartt committed:
Merge branch 'feat/2615-instructions-task' of github.com:argilla-io/argilla into feat/list-datasets-client

alvarobartt committed:
Remove return type-hint `List[FeedbackDataset]`

alvarobartt committed:
Update `list_datasets` return type-hint

alvarobartt committed:
Update `regex` for `BaseDatasetModel.name`

We can have names with spaces and upper case letters e.g. `Big Dataset` is a valid dataset name, and the current validation fails in those cases

alvarobartt committed:
Remove `list_datasets` import in `__init__.py`

alvarobartt committed:
Fix `BaseDatasetModel` to keep previous plus spaces

alvarobartt committed:
Call `check_schema_props` just on Python `dict`s

alvarobartt committed:
Add `list_datasets` in `_import_structure`

alvarobartt committed:
Remove `list_datasets` from `__init__` & `client/api.py`

alvarobartt committed:
Keep default `regex` in `BaseDatasetModel.name`

To match the server's default `regex` for validating the dataset names, we are keeping it as is

alvarobartt committed:
Add `/api/v1` endpoints in a separate module

alvarobartt committed:
Add `Dataset` class

For the moment just working with the `/api/v1` and intended to be used for the `FeedbackTask` only, until properly defined

alvarobartt committed:
Add `Dataset` class in `__init__`

To be imported/used as `import argilla as rg; dataset = rg.Dataset(...)`

alvarobartt committed:
Fix missing `Union` type-hint

alvarobartt committed:
Infer `schema` when adding first record

alvarobartt commented on src/argilla/client/sdk/datasets/v1/models.py:
```diff
@@ -0,0 +1,34 @@
+class BaseDatasetModel(BaseModel):
+    tags: Dict[str, str] = Field(default_factory=dict)
+    metadata: Dict[str, Any] = Field(default_factory=dict)
+    name: str = Field(regex="^(?!-|_)[a-zA-Z0-9-_ ]+$")
```
@jfcalvo I saw that datasets can be created using spaces, which was not supported before, is this correct?

jfcalvo replied:
That's right. We don't have any limitation about what a dataset name can be. We will have soon a limitation in the number of characters but that's all.

jfcalvo replied:
With that said maybe you don't need to specify any regex. That should be transparent to you.

alvarobartt replied:
Since I was validating the schema of the dataset uploaded to Argilla, I changed that since the dataset I was using was named `"Big Dataset"` and the validation was failing, but we can make it just a string with no regex validation.

alvarobartt commented on src/argilla/client/sdk/datasets/v1/api.py:
```diff
@@ -0,0 +1,152 @@
+    return handle_response_error(response)
+
+
+def add_record(
```
Also it seems that we're no longer able to add records until the dataset is published, is this correct? @jfcalvo

jfcalvo replied:
Exactly. Right now the flow should be the following:
1. Create the Dataset.
2. Add at least one Field to the Dataset.
3. Add at least one Question to the Dataset.
4. Publish the Dataset.
5. Add records to the dataset that matches the defined fields and questions for the dataset.

Before publishing the dataset is not possible to add records.

alvarobartt replied:
Cool, I'll push both the `fields` and `questions` methods today to cover all the `/api/v1` endpoints 😄

alvarobartt commented on src/argilla/client/sdk/datasets/models.py:
```diff
@@ -36,7 +36,7 @@ def _missing_(cls, value):
     tags: Dict[str, str] = Field(default_factory=dict)
     metadata: Dict[str, Any] = Field(default_factory=dict)
-    name: str = Field(regex="^(?!-|_)[a-z0-9-_]+$")
+    name: str = Field(regex="^(?!-|_)[a-zA-Z0-9-_]+$")
```
I also updated this regex, but didn't include the space just in case

jfcalvo replied:
If that's the old Dataset model I think we should not change it. Correct me if I'm wrong @frascuchon

alvarobartt replied:
Makes sense, but was that intended originally? I mean, not allowing upper-case letters in the dataset name

jfcalvo replied:
The idea was to format the dataset name so it can be used as ID for ElasticSearch indexes.

alvarobartt commented on src/argilla/client/nightly.py:
```diff
@@ -0,0 +1,231 @@
+        return self.__records[0]
+
+    # TODO: we could fetch those on iter, maybe we can create an `streaming` flag or something similar
+    def iter(self, batch_size: int = 32) -> List[BaseModel]:
```
Here I'll probably create another class named `StreamingDataset` to stream the data during training if that's something we need/want, but still need to discuss it

alvarobartt commented on src/argilla/client/nightly.py:
```diff
@@ -0,0 +1,231 @@
+    settings: Dict[str, Any]
+
+
+class Dataset:
```
I'm planning to combine the current with some ideas that @frascuchon left at https://gist.github.com/frascuchon/2e11db8a00559746e4b4c2dff687daaf

alvarobartt committed:
Add `__getitem__` and `__len__`

alvarobartt committed:
Show `warnings.warn` if `Dataset` already published

alvarobartt committed:
Fix `add_records` call since limit needs to be fixed

Currently defaults to 1 to run experiments, but we should increase it to at least 250 (current default chunk size in `rg.load`)

alvarobartt committed:
Add `batch` to `Dataset.records` to cache it

alvarobartt committed:
Add `create_dataset` from `/api/v1`

alvarobartt committed:
Add `__del__`, `__enter__` & `__exit__` to `Dataset`

alvarobartt committed:
Add `/me` to current user's endpoints

alvarobartt committed:
Add `fields` & `questions` requests/methods

alvarobartt committed:
Add `create_feedback_dataset` & `FeedbackDataset`

alvarobartt committed:
Define `rg.create_feedback_dataset` & `rg.FeedbackDataset`

alvarobartt committed:
Fix `fields` & `questions` properties return type-hint

alvarobartt committed:
Fix `FeedbackDatasetModel` as of `/api/v1`

alvarobartt committed:
Add `RecordsModel` schema

alvarobartt committed:
Update `/api/v1/datasets` return type-hints

alvarobartt committed:
Add `list_workspaces` to list all the available workspaces

Required for the `/api/v1` endpoints that expect the `workspace_id` as the input, since the user doesn't know the ID, but the name

alvarobartt committed:
Add `Workspace` class to get ID from name

alvarobartt committed:
Add `Workspace` class to top-level imports

alvarobartt committed:
Add `list_workspaces` & `Workspace` docstrings

alvarobartt committed:
Add `list_workspaces` & `Workspace` unit tests

alvarobartt committed:
Add `from_name` classmethod in `rg.Workspace`

Co-authored-by: Francisco Aranda <francis@argilla.io>

alvarobartt committed:
Add `Workspace` class & `list_workspaces` (#2859)

# Description

This PR has been created on top of #2827 to add a `Workspace` class so
that we can easily convert workspace names (human readable format) to
workspace IDs (transparent to the user). This is just covering the basic
functionality which is to `list_workspaces` to get all the available
workspaces for the current user, so that when instantiating `Workspace`
we can know whether that workspace exists and is available for the
current user or not.

Besides that, the main reason behind this PR is because the `/api/v1`
and `/api/v1/me` endpoints expect the workspace ID, not the workspace
name, as opposed to the former `/api` (v0) which was using names not
IDs.

> :warning: Note that this may be extended in the future to cover all
the `Workspace` functionalities provided by Argilla, but for the moment,
we're keeping it simple since it's just going to be used for name to ID
conversation and workspace validation.

**Type of change**

- [X] New feature (non-breaking change which adds functionality)

**How Has This Been Tested**

- [X] Add unit tests for `list_datasets` and `Workspace`

**Checklist**

- [X] I have merged the original branch into my forked branch
- [X] I added relevant documentation
- [X] follows the style guidelines of this project
- [X] I did a self-review of my code
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my
feature works

alvarobartt committed:
Append `record` to `self.__records`

alvarobartt committed:
Use access through `pydantic` schema instead of `dict`

alvarobartt committed:
Rename to `FeedbackDataset` & update schema

Due to the recent rename of the `rg.Dataset` to `rg.FeedbackDataset`

alvarobartt committed:
Use `rg.Workspace` in `rg.FeedbackDataset`

alvarobartt committed:
Update `pydantic.BaseModel` schemas

alvarobartt committed:
Include `responses` by default in `OnlineRecordSchema`

alvarobartt committed:
Add `create_feedback_dataset` function

alvarobartt committed:
Add `delete_dataset` method via `/api/v1`

alvarobartt committed:
Remove unused print-statement

alvarobartt committed:
Handle exceptions in `create_feedback_dataset`

alvarobartt committed:
Add `TextField`, `TextQuestion`, & `RatingQuestion` as top-level imports

alvarobartt committed:
Remove unused methods

alvarobartt committed:
Add `FeedbackRecordsModel`, `FeedbackItemModel` & `FeedbackResponseModel` schemas

alvarobartt committed:
Use `FeedbackRecordsModel` & update type-hints

alvarobartt committed:
Add `FeedbackRecord` schema

alvarobartt committed:
Remove unused `pydantic.BaseModel` schemas

alvarobartt committed:
Use `FeedbackItemModel` schema

alvarobartt committed:
Add `add_records` in replacement of `add_record`

To be aligned with the API we should be able to log multiple records at once.

alvarobartt committed:
Make `black` and `ruff` smile again

alvarobartt committed:
Fix `typing.Literal` imports on Python 3.7

frascuchon commented on src/argilla/client/client.py:
```diff
@@ -208,6 +215,50 @@ def get_workspace(self) -> str:
+        return datasets_api_v1.get_questions(client=self._client, id=id).parsed
+
+    def add_question(self, id: str, question: Dict[str, Any]) -> httpx.Response:
+        return datasets_api_v1.add_question(client=self._client, id=id, question=question)
```
All these methods are referring to the new apiV1 endpoints and could be confusing to keep them here as is. For example, the `list_datasets` method will return only new kinds of datasets. 

Could we move them to a DatasetsV1 class and work with them from there?

```python 
class DatasetsV1:

   def __init__(client: AuthenticatedCliewnt):
       self.client = client

  def list_datasets(self) -> List[FeedbackDatasetModel]:
        return datasets_api_v1.list_datasets(client=self._client).parsed
```

And then use as:

```python
argilla_client.datasets_v1.list_datasets()
```

alvarobartt replied:
Yes, you're right! I'll update this too, thanks 👍🏻

frascuchon commented on src/argilla/client/nightly.py:
If this this module defines the user layer data models, we should rename it to `client/feedback.py` and work from that

alvarobartt replied:
Yes, I was keeping the nightly just until we decided this, so I'll rename it now!

frascuchon commented on src/argilla/client/nightly.py:
```diff
@@ -0,0 +1,367 @@
+        workspace: Optional[Union[rg.Workspace, str]] = None,
+        id: Optional[str] = None,
+    ) -> None:
+        self.client: "Argilla" = rg.active_client()
```
I prefer to let users work locally with these feedback datasets, and do explicit interactions with the argilla server, as `dataset.save` or `dataset.push`

Similar for loading. We should load datasets from argilla with an explicit method.
````
python
ds = FeedbackDataset.from_name_and_workspace(name, workspace)
...
````

alvarobartt replied:
Yes, I was already preparing `push_to_argilla` and `from_argilla` as we discussed! 😄

frascuchon commented on src/argilla/client/nightly.py:
```diff
@@ -0,0 +1,367 @@
+
+    if workspace is None or isinstance(workspace, str):
+        try:
+            workspace = rg.Workspace.from_name(workspace)
```
If the `workspace` is `None` this statement should fail. Could we use the `client.get_current_workspace()` to setup a default value?
```python
workspace = workspace or client.get_current_workspace()
```

alvarobartt replied:
Indeed the `rg.Workspace.from_name` uses the default if the provided is None, so it won't fail 👍🏻

alvarobartt replied:
Unless you prefer to set `name` in `rg.Workspace.from_name` as a mandatory value, and get the default in the line you mentioned instead.

frascuchon commented on src/argilla/client/sdk/datasets/models.py:
```diff
@@ -36,7 +36,7 @@ def _missing_(cls, value):
     tags: Dict[str, str] = Field(default_factory=dict)
     metadata: Dict[str, Any] = Field(default_factory=dict)
-    name: str = Field(regex="^(?!-|_)[a-z0-9-_]+$")
+    name: str = Field(regex="^(?!-|_)[a-zA-Z0-9-_]+$")
```
For old datasets this is not a valid name (uppercase letters are not allowed), we should revert this change here

alvarobartt replied:
You're right! I'll rever the `regex` validation update there!

frascuchon commented on src/argilla/client/sdk/datasets/v1/__init__.py:
I think it makes sense changing a bit the module structure.  Instead of 
````
sdk/
  datasets/
    v1/
````
we can group by sdk API versions (at least for the v1 one):
````
sdk/
  v1/
    datasets/
````

alvarobartt replied:
Sure, I'll change this rn!

frascuchon commented on src/argilla/client/sdk/datasets/v1/api.py:
```diff
@@ -0,0 +1,301 @@
+    return handle_response_error(response)
+
+
+@lru_cache(maxsize=None)
```
we can remove this for now, and introduce a cache mechanism later. This method can be called several times and invalidating the cache can be a hell

alvarobartt replied:
Sure, I used it because it was already used in the former `datasets/api.py`

alvarobartt committed:
Revert `regex` to allow upper-case letters

Co-authored-by: Francisco Aranda <francis@argilla.io>

alvarobartt committed:
Rename `nightly` to `feedback`

Co-authored-by: Francisco Aranda <francis@argilla.io>

alvarobartt committed:
Remove `lru_cache` usage temporarily

Co-authored-by: Francisco Aranda <francis@argilla.io>

alvarobartt committed:
Move `sdk/datasets/v1` to `sdk/v1/datasets`

Co-authored-by: Francisco Aranda <francis@argilla.io>

alvarobartt committed:
Merge branch 'feat/2615-instructions-task' into feat/new-datasets

alvarobartt committed:
Fix `ValidationApiError` due to missing `params`

alvarobartt committed:
Use `StrictStr` and `StrictInt`

Done to avoid parsing `int` as `str`

alvarobartt committed:
Add `FeedbackFieldModel` & `FeedbackQuestionModel`

alvarobartt committed:
Use `construct` since data is validated at API-level

alvarobartt committed:
Fix to exclude `FeedbackRecord` fields if `None`

alvarobartt committed:
Import `typing_extensions.Literal` on Python 3.7 or below

alvarobartt committed:
Rename `schema` to `fields_schema`

alvarobartt committed:
Add `FeedbackValueModel` to validate `values`

alvarobartt committed:
Add `OfflineFeedback...` schemas

alvarobartt committed:
Add `fetch_records` & set `FETCHING_BATCH_SIZE`

Previously those were retrieved when accessing the property `records`, but now are fetched using `fetch_records`

alvarobartt committed:
Refactor `FeedbackDataset` to use locally

alvarobartt committed:
Add `push_to_argilla` & `from_argilla`

alvarobartt committed:
Clean duplicated code

alvarobartt committed:
Fix `add_records` type-hints

alvarobartt committed:
Add `FeedbackDataset.iter` back

alvarobartt committed:
Remove `loaded_from_argilla` & add `argilla_id`

alvarobartt committed:
Push to records to Argilla if `push=True`

alvarobartt committed:
Add records on `push_to_argilla` if applicable

alvarobartt committed:
Minor changes

alvarobartt committed:
Add `lru_cache` back for `get_dataset`

CI was failing because we were calling `get_dataset.cache_clear()` on a function that had no `lru_cache`

frascuchon commented on src/argilla/client/feedback.py:
```diff
@@ -0,0 +1,454 @@
+    def __getitem__(self, key: Union[slice, int]) -> Union["FeedbackItemModel", List["FeedbackItemModel"]]:
+        if len(self.__records) < 1:
+            raise RuntimeError(
+                "In order to get items from `rg.FeedbackDataset` you need to either add them first with `add_records` or fetch them from Argilla or HuggingFace with `fetch_records`."
```
```suggestion
                "In order to get items from `rg.FeedbackDataset` you need to either add them first with `add_records` "
                "or fetch them from Argilla or HuggingFace with `fetch_records`."
```

frascuchon commented on src/argilla/client/feedback.py:
```diff
@@ -0,0 +1,454 @@
+    def records(self) -> List["FeedbackItemModel"]:
+        return self.__records
+
+    def fetch_records(self, overwrite: bool = False) -> List["FeedbackItemModel"]:
```
So users can refresh info from argilla using this method. How users can fetch a chunk of data `ds.records[100:1000]` ?

alvarobartt replied:
The thing is that the only consistent solution is to fetch all the records from Argilla when calling `from_argilla`, as otherwise relying on the API for every single record will end up on a consistency issue... We can discuss this later today with a demo!

frascuchon commented on src/argilla/client/feedback.py:
```diff
@@ -0,0 +1,454 @@
+                    batch = client.get_records(
+                        id=self.argilla_id, offset=FETCHING_BATCH_SIZE * current_batch, limit=FETCHING_BATCH_SIZE
+                    )
+                    self.__records += batch.items
```
If we do:

```python
ds.fetch_records(override=True)
```
and later 
```python
ds.fetch_records(override=False) # <- warning here!
```

This will show a message that can be confusing regarding the non-pushed records (which are records fetched previously)

alvarobartt committed:
Merge branch 'feat/2615-instructions-task' into feat/new-datasets

alvarobartt committed:
Update src/argilla/client/feedback.py

Co-authored-by: Francisco Aranda <francis@argilla.io>

alvarobartt committed:
Merge branch 'feat/2615-instructions-task' into feat/new-datasets

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks
for more information, see https://pre-commit.ci

alvarobartt committed:
Make `black` happy

alvarobartt committed:
Improve record `push` and `fetch` mechanism

alvarobartt committed:
Merge branch 'feat/new-datasets' of github.com:argilla-io/argilla into feat/new-datasets

alvarobartt committed:
Construct `fields` & `questions` from Python dicts

alvarobartt committed:
Use `self.fields` in `generate_pydantic_schema`

tomaarsen commented on src/argilla/client/sdk/v1/datasets/models.py:
```diff
@@ -0,0 +1,87 @@
+if sys.version_info >= (3, 8):
+    from typing import Literal
+else:
+    from typing_extensions import Literal
```
The most robust solution may be to use the Python [EAFP](https://docs.python.org/3/glossary.html#term-EAFP) coding style:
```python
try:
    from typing import Literal
except ImportError:
    from typing_extensions import Literal
```
That way, you'll never accidentally list the wrong `sys` version, and it won't break if in the future the Python devs decide to moving something from `typing` back to `typing_extensions` or something. (I assume they won't, but still).

This extends to the rest of the codebase here. I'm open to your thoughts here.

alvarobartt replied:
Well since that certainly would work, I guess that if we can catch it better right? Anyway it's a minor issue so I'm also open to catch that on `ImportError` and fallback to `typing_extensions` if `typing` doesn't work for a certain Python version e.g. `typing.Annotated` introduced in Python 3.9, or `typing.Literal` introduced in Python 3.8.

alvarobartt committed:
Fix `create_dataset` type-hint

alvarobartt committed:
Add missing docstrings

frascuchon committed:
Using an already setup httpx client in sdk modules

frascuchon committed:
fixing tests

alvarobartt committed:
Fix `FeedbackItemModel` schema type-hint

alvarobartt committed:
Add `FeedbackDataset` unit tests (WIP)

alvarobartt committed:
Forbid extra fields in `rg.FeedbackRecord`

alvarobartt committed:
Fix `rg.FeedbackRecord.responses` validation

alvarobartt committed:
Use valid `pytest.fixtures`

alvarobartt committed:
Fix default URL in docstring

alvarobartt committed:
Add `/api/v1` tests for `FeedbackDataset` (WIP)

alvarobartt committed:
Fallback to `typing_extensions` on `ImportError`

Co-authored-by: Tom Aarsen <37621491+tomaarsen@users.noreply.github.com>

alvarobartt committed:
Fix `push_to_argilla`

alvarobartt committed:
Fix `test_push_to_argilla_and_from_argilla` tests

alvarobartt committed:
Using an already setup httpx client in sdk modules (#2919)

# Description

Since the new SDK modules are creating a new httpx client on each
request, some problems related to connection performance can be raised.
In the [official docs](https://www.python-httpx.org/advanced/) you can
read:

> When you make requests using the top-level API as documented in the
[Quickstart](https://www.python-httpx.org/quickstart/) guide, HTTPX has
to establish a new connection for every single request (connections are
not reused). As the number of requests to a host increases, this quickly
becomes inefficient.
> 
> On the other hand, a Client instance uses [HTTP connection
pooling](https://en.wikipedia.org/wiki/HTTP_persistent_connection). This
means that when you make several requests to the same host, the Client
will reuse the underlying TCP connection, instead of recreating one for
every single request.

In order to better use the httpx client, this PR changes the `sdk`
module to reuse the already set up httpx client in the
`AuthenticatedClient` instance.

By using this client, you don't need to pass header or base URL
configuration on each request, since is setup on client initialization.

alvarobartt committed:
Fix minor typo (requet -> request)

alvarobartt committed:
Fix failing tests due to `httpx.Client`

alvarobartt committed:
Rename `test_feeedback` to `test_feedback`

alvarobartt committed:
Add missing `/api/v1/datasets` docstrings

alvarobartt committed:
Add every `/api/v1/datasets` unit test

alvarobartt committed:
Remove `__future__.annotations` as no longer required

Co-authored-by: frascuchon <francis@argilla.io>

frascuchon commented on src/argilla/client/feedback.py:
```diff
@@ -0,0 +1,809 @@
+    settings: Dict[str, Any] = Field({"type": "text"}, const=True)
+
+
+class RatingQuestion(QuestionSchema):
+    settings: Dict[str, Any] = Field({"type": "rating"})
+    values: List[int]
+
+    @validator("values", always=True)
+    def update_settings_with_values(cls, v, values):
+        if v:
+            values["settings"]["options"] = [{"value": value} for value in v]
+        return v
+
```
If we want to keep updated values and settings once the question object is created, we should enable the `validate_assignment` config attribute. See [docs](https://docs.pydantic.dev/latest/usage/model_config/#options) 
```suggestion
class RatingQuestion(QuestionSchema):
    settings: Dict[str, Any] = Field({"type": "rating"})
    values: List[int]

    @validator("values", always=True)
    def update_settings_with_values(cls, v, values):
        if v:
            values["settings"]["options"] = [{"value": value} for value in v]
        return v
    
    class Config:
        validate_assignment= True
```

alvarobartt replied:
You mean to do something like e.g.

```python
import argilla as rg

question = rg.RatingQuestion(name="rating", required=True, values=[1, 2, 3])
question.values = [1, 2, 3, 4, 5]
```

Not sure if that's what you mean

frascuchon replied:
Exactly

alvarobartt replied:
Ok then I'll update that!

frascuchon commented on src/argilla/client/feedback.py:
```diff
@@ -0,0 +1,807 @@
+        if hasattr(self, "__new_records") and self.__new_records is not None:
+            del self.__new_records
+
+    def __enter__(self) -> "FeedbackDataset":
```
What the use case you're thinking about the context manager?

alvarobartt replied:
I used it a couple times while testing everything, but probably there's no use-cases, so we can remove this

alvarobartt committed:
Add double-quotes when type-checking enabled

frascuchon commented on src/argilla/client/feedback.py:
```diff
@@ -0,0 +1,807 @@
+                total=total_batches,
+                desc="Fetching records from Argilla",
+            ) as pbar:
+                while current_batch <= total_batches:
```
Since the `total` is optional (See the [API ref](https://tqdm.github.io/docs/tqdm/#__init__)), we still use the tqdm progress bar and prepare changes for #2917

```suggestion
            current_batch = 1
            with tqdm(
                initial=current_batch,
                desc="Fetching records from Argilla",
            ) as pbar:
                while current_batch <= total_batches:
```

alvarobartt replied:
Then I'll need to change the whole loop I guess, since I was also using the computed `total_batches` for updating the `offset` on every request to Argilla API

alvarobartt replied:
Is there any way to retrieve the total number of records from the metrics?

frascuchon commented on src/argilla/client/client.py:
```diff
@@ -208,6 +217,143 @@ def get_workspace(self) -> str:
+        return datasets_api_v1.create_dataset(
+            client=self._client.httpx,
+            name=name,
+            workspace_id=workspace_id,
```
Since all these methods are referring the new FeedbackDataset and are used in the `feedback.py`, I would move to the module as private functions, for 2 main reasons:

1. Keep current/old code as invariant as possible -> (new things in a different place will help us with the migration progress)
2. *Dataset* is a concept already defined in Argilla, so can be confusing having a `list_datasets` method returning only V1 ones.

alvarobartt replied:
Yes, I remember you mentioned this some days ago, can we schedule a quick call to see the approach you're thinking of? Thanks 😄

alvarobartt committed:
Add `validate_assignment` in `RatingQuestion`

Co-authored-by: frascuchon <francis@argilla.io>

alvarobartt committed:
Remove `total` usage since it will be removed

Co-authored-by: frascuchon <francis@argilla.io>

alvarobartt committed:
Remove `/api/v1/datasets` wrapper from `client.py`

frascuchon committed:
Merge branch 'feat/2615-instructions-task' into feat/new-datasets

alvarobartt committed:
Fix unit tests since `len(guidelines)` must be >1

alvarobartt committed:
Add `guidelines` length-validation
